File: Adramelech/Commands/Slash/Weather.cs
using System.Diagnostics.CodeAnalysis;
using System.Text.Json.Serialization;
using Adramelech.Configuration;
using Adramelech.Extensions;
using Adramelech.Services;
using Adramelech.Utilities;
using Discord;
using Discord.Interactions;
using Discord.WebSocket;
using Flurl;

namespace Adramelech.Commands.Slash;

public class Weather(Config config, HttpUtils httpUtils, CooldownService cooldownService)
    : InteractionModuleBase<SocketInteractionContext<SocketSlashCommand>>
{
    private const string OpenWeatherUrl = "https://api.openweathermap.org";

    [SlashCommand("weather", "Get the weather of a city")]
    public async Task WeatherAsync([Summary("country", "The country of the city")] string country,
        [Summary("city", "The city to get the weather of")]
        string city)
    {
        if (await Context.VerifyCooldown(cooldownService)) return;
        if (config.FeedbackWebhook.IsNullOrEmpty())
        {
            await Context.SendError("The feedback webhook is not set up.");
            return;
        }

        await DeferAsync();

        var coordinates = await httpUtils.GetAsync<OpenWeatherGeo[]>(
            OpenWeatherUrl.AppendPathSegments("geo", "1.0", "direct")
                .SetQueryParam("q", $"{city},{country}")
                .SetQueryParam("appid", config.OpenWeatherKey)
                .ToString());
        if (coordinates.IsDefault()) await Context.SendError("Failed to get the coordinates of the city.");

        var weather = await httpUtils.GetAsync<OpenWeather>(
            OpenWeatherUrl.AppendPathSegments("data", "2.5", "weather")
                .SetQueryParam("lat", coordinates![0].Lat)
                .SetQueryParam("lon", coordinates[0].Lon)
                .SetQueryParam("appid", config.OpenWeatherKey)
                .SetQueryParam("units", "metric")
                .SetQueryParam("lang", "en")
                .ToString());
        if (weather.IsDefault())
        {
            await Context.SendError("Failed to get the weather of the city.");
            return;
        }

        var mainField = $"""
                         **Temperature**: {weather.Main.Temp}°C
                         **Feels Like**: {weather.Main.FeelsLike}°C
                         **Minimum Temperature**: {weather.Main.TempMin}°C
                         **Maximum Temperature**: {weather.Main.TempMax}°C
                         **Pressure**: {weather.Main.Pressure}hPa
                         **Humidity**: {weather.Main.Humidity}%
                         **Sea Level**: {weather.Main.SeaLevel}hPa
                         **Ground Level**: {weather.Main.GroundLevel}hPa
                         """;
        var weatherField = $"""
                            **Main**: {weather.Weather[0].Main}
                            **Description**: {weather.Weather[0].Description}
                            """;
        var windField = $"""
                         **Speed**: {weather.Wind.Speed}m/s
                         **Degree**: {weather.Wind.Deg}°
                         **Gust**: {weather.Wind.Gust}m/s
                         """;

        await FollowupAsync(embed: new EmbedBuilder()
            .WithColor(config.EmbedColor)
            .WithTitle($"Weather{(weather.Name.IsNullOrEmpty() ? string.Empty : $" in {weather.Name}")}")
            .AddField("> :zap: Main", mainField)
            .AddField("> :cloud: Weather", weatherField)
            .AddField("> :dash: Wind", windField)
            .WithFooter("Powered by openweathermap.org")
            .Build());
        Context.SetCooldown(cooldownService, TimeSpan.FromMinutes(10));
    }

    [SuppressMessage("ReSharper", "UnusedAutoPropertyAccessor.Local")]
    private struct OpenWeatherGeo
    {
        public double Lat { get; set; }
        public double Lon { get; set; }
    }

    [SuppressMessage("ReSharper", "UnusedAutoPropertyAccessor.Local")]
    private struct OpenWeather
    {
        public SWeather[] Weather { get; set; }
        public SMain Main { get; set; }
        public SWind Wind { get; set; }
        public string? Name { get; set; }

        internal struct SWeather
        {
            public string Main { get; set; }
            public string Description { get; set; }
        }

        internal struct SMain
        {
            public double Temp { get; set; }
            public double FeelsLike { get; set; }
            public double TempMin { get; set; }
            public double TempMax { get; set; }
            public int Pressure { get; set; }
            public int Humidity { get; set; }
            public int SeaLevel { get; set; }
            [JsonPropertyName("grnd_level")] public int GroundLevel { get; set; }
        }

        internal struct SWind
        {
            public double Speed { get; set; }
            public int Deg { get; set; }
            public double Gust { get; set; }
        }
    }
}

File: src/bot/commands/slash/weather.ts
import { stripIndents } from 'common-tags';
import {
  ChatInputCommandInteraction,
  ComponentType,
  MessageFlags,
  SlashCommandBuilder,
} from 'discord.js';
import ky from 'ky';
import v from 'voca';
import { z } from 'zod';
import env from '#env';
import type { Command } from '#bot/types/command';
import { sendError } from '#bot/utils/sendError';

const BASE_URL = 'https://api.openweathermap.org';

const geoSchema = z.object({
  lat: z.number(),
  lon: z.number(),
});

const weatherSchema = z.object({
  name: z.string().nullish(),
  weather: z.array(
    z.object({
      main: z.string(),
      description: z.string(),
    })
  ),
  main: z.object({
    temp: z.number(),
    feels_like: z.number(),
    temp_min: z.number(),
    temp_max: z.number(),
    pressure: z.number(),
    humidity: z.number(),
    sea_level: z.number(),
    grnd_level: z.number(),
  }),
  wind: z.object({
    speed: z.number(),
    deg: z.number(),
    gust: z.number().optional(), // Present most of the time, but not always
  }),
});

export const command = <Command>{
  data: new SlashCommandBuilder()
    .setName('weather')
    .setDescription('Get the weather of a city')
    .addStringOption((option) =>
      option
        .setName('city')
        .setDescription('The city you want to get the weather of')
        .setRequired(true)
    )
    .addStringOption((option) =>
      option
        .setName('country')
        .setDescription('The country of the city')
        .setRequired(true)
    ),
  cooldown: 10 * 60, // 10 minutes, because the API is expensive and I'm not rich
  uses: ['OpenWeatherMap'],
  async execute(intr: ChatInputCommandInteraction) {
    if (!env.OPENWEATHER_KEY)
      return await sendError(intr, 'OpenWeatherMap API key is not configured');

    await intr.deferReply();

    const city = intr.options.getString('city', true);
    const country = intr.options.getString('country', true);

    const geoUrl = new URL(`${BASE_URL}/geo/1.0/direct`);
    geoUrl.searchParams.append('q', `${city},${country}`);
    geoUrl.searchParams.append('limit', '1');
    geoUrl.searchParams.append('appid', env.OPENWEATHER_KEY);

    const geoResponse = await ky(geoUrl.toString()).json();
    const { data: geoData, error: geoError } = geoSchema
      .array()
      .safeParse(geoResponse);
    if (geoError) return await sendError(intr, 'City not found');

    const weatherUrl = new URL(`${BASE_URL}/data/2.5/weather`);
    weatherUrl.searchParams.append('lat', geoData[0].lat.toString());
    weatherUrl.searchParams.append('lon', geoData[0].lon.toString());
    weatherUrl.searchParams.append('appid', env.OPENWEATHER_KEY);
    weatherUrl.searchParams.append('units', 'metric');
    weatherUrl.searchParams.append('lang', 'en');
    const response = await ky(weatherUrl.toString()).json();
    const { data, error } = weatherSchema.safeParse(response);
    if (error)
      return await sendError(
        intr,
        'An error occurred while fetching the weather data'
      );

    await intr.followUp({
      flags: MessageFlags.IsComponentsV2,
      components: [
        {
          type: ComponentType.Container,
          accent_color: env.EMBED_COLOR,
          components: [
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              # Weather${data.name ? ` in ${data.name}` : ''}
              **Temperature:** ${data.main.temp}°C
              **Feels like:** ${data.main.feels_like}°C
              **Minimum temperature:** ${data.main.temp_min}°C
              **Maximum temperature:** ${data.main.temp_max}°C
              **Pressure:** ${data.main.pressure} hPa
              **Humidity:** ${data.main.humidity}%
              **Sea level:** ${data.main.sea_level} hPa
              **Ground level:** ${data.main.grnd_level} hPa
              ### :cloud: Weather
              **Main:** ${data.weather[0].main}
              **Description:** ${v.titleCase(data.weather[0].description)}
              ### :dash: Wind
              **Speed:** ${data.wind.speed} m/s
              **Direction:** ${data.wind.deg}°
              **Gust:** ${data.wind.gust ? `${data.wind.gust} m/s` : 'N/A'}
              `,
            },
            {
              type: ComponentType.TextDisplay,
              content: '> Powered by OpenWeatherMap',
            },
          ],
        },
      ],
    });
  },
};
